### PR TITLE
Backbone Hidden Noise: Gaussian noise injection for OOD robustness

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Round 19's spectral norm experiment (#2243) revealed a key insight: **OOD failure is in the backbone representation, not the output heads.** Spectral norm on SRF heads improved p_in (-2.5%) but worsened OOD metrics — because the backbone features themselves are erratic on OOD inputs.

**Adding small Gaussian noise to backbone hidden states during training** directly addresses this. By perturbing the learned features at each TransolverBlock, we force the downstream layers (and SRF heads) to be robust to small representation perturbations — which is exactly what OOD inputs cause.

This is fundamentally different from:
- **Input noise** (#2235, failed): corrupts geometric features before encoding
- **Dropout** (#2245, failed): removes entire neurons, too aggressive for SRF
- **Weight noise / SAM**: perturbs parameters, not features

**Backbone hidden noise** adds noise at the RIGHT level of abstraction — after encoding but before output — forcing the model to learn smooth, robust mappings from features to predictions.

**Analogies:**
- Denoising autoencoders (Vincent et al., 2008): inject noise in latent space, force reconstruction robustness
- Noise injection in RNNs (Gal & Ghahramani, 2016): improves sequence model robustness
- Gaussian noise regularization (Bishop, 1995): equivalent to Tikhonov regularization on the learned representation

**σ = 0.01**: Very small relative to hidden state magnitudes (which are O(1) after LayerNorm). Enough to regularize without destroying signal.

**Confidence:** Medium. The theoretical motivation is strong (directly targets the identified failure mode). The uncertainty is in calibrating σ — too large destroys convergence, too small has no effect.

### Code changes in `cfd_tandemfoil/train.py`

**Step 1: Add config flag**

In the `Config` class, add:
```python
backbone_noise: float = 0.0   # std of Gaussian noise added to backbone hidden states during training (0=disabled)
```

**Step 2: Add noise in TransolverBlock.forward()**

In the `TransolverBlock.forward()` method, add noise to the output hidden state AFTER the residual connection and LayerNorm but BEFORE returning. This way, each block's output is perturbed:

```python
def forward(self, fx, ...):
    # ... existing forward pass ...
    # At the very end, after all computations:
    if self.training and cfg.backbone_noise > 0:
        fx = fx + torch.randn_like(fx) * cfg.backbone_noise
    return fx
```

**IMPORTANT:** 
- Only during training (`self.training`), never at inference.
- Apply AFTER the residual connection — we want to perturb the COMBINED representation, not just the attention output.
- Apply to ALL TransolverBlocks (the noise compounds slightly across layers, which is fine at σ=0.01).

**Step 3: Verify torch.compile compatibility**

`torch.randn_like` should be compatible with torch.compile. If not, use a pre-allocated noise buffer.

### Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py \
  --agent edward --wandb_name "edward/backbone-noise-s42" \
  --wandb_group "round20/backbone-hidden-noise" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --backbone_noise 0.01

# Seed 73
cd cfd_tandemfoil && python train.py \
  --agent edward --wandb_name "edward/backbone-noise-s73" \
  --wandb_group "round20/backbone-hidden-noise" \
  --seed 73 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --backbone_noise 0.01
```

New flag: `--backbone_noise 0.01`

### Report results

Table: p_in, p_oodc, p_tan, p_re for both seeds, 2-seed avg, baseline comparison, W&B run IDs.

**Focus on p_oodc and p_re** — these are the OOD metrics that should benefit most from backbone robustness. If p_in degrades, the noise may be too large. If nothing changes, the noise may be too small.

## Baseline

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in   | **11.979** | < 11.98 |
| p_oodc | **7.643**  | < 7.65  |
| p_tan  | **28.341** | < 28.34 |
| p_re   | **6.300**  | < 6.30  |

W&B: `hgml7i2r` (s42), `qic03vrg` (s73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent edward --wandb_name "edward/baseline" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT edward:

## Results

### Backbone Hidden Noise (σ=0.01)

Added Gaussian noise injection to TransolverBlock hidden states during training. Noise is added after squeeze-excitation + FiLM conditioning, before last-layer heads, in all 3 TransolverBlocks. Uses `torch.randn_like(fx) * cfg.backbone_noise`, gated by `self.training`.

### Metrics

| Metric | Baseline | Seed 42 | Seed 73 | 2-seed Avg | Δ vs Baseline |
|--------|----------|---------|---------|------------|---------------|
| p_in   | **11.979** | 12.216 | 12.000 | **12.108** | +0.129 ❌ |
| p_oodc | **7.643**  | 8.0    | 7.6    | **7.800**  | +0.157 ❌ |
| p_tan  | **28.341** | 28.7   | 28.2   | **28.450** | +0.109 ❌ |
| p_re   | **6.300**  | 6.7    | 6.4    | **6.550**  | +0.250 ❌ |

**Val loss:** s42=0.3803 (epoch 146), s73=0.3753 (epoch 145)

### Surface MAE (in-dist, from W&B)

| Component | Seed 42 | Seed 73 |
|-----------|---------|---------|
| Ux        | 2.563   | 2.621   |
| Uy        | 0.844   | 0.853   |
| p         | 12.216  | 12.000  |

### Volume MAE (in-dist, from W&B)

| Component | Seed 42 | Seed 73 |
|-----------|---------|---------|
| Ux        | 0.515   | 0.516   |
| Uy        | 0.191   | 0.197   |
| p         | 12.587  | 12.797  |

### Commands

```bash
# Seed 42
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent edward --wandb_name "edward/backbone-noise-s42" \
  --wandb_group "round20/backbone-hidden-noise" --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature --backbone_noise 0.01

# Seed 73 — identical but --seed 73 --wandb_name "edward/backbone-noise-s73"
```

### W&B Runs
- Seed 42: `qhex5dph` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/qhex5dph
- Seed 73: `sq588kij` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/sq588kij

### Peak Memory
Both runs trained on single H100 GPUs. No OOM issues.

### What happened

**All metrics regressed. The hypothesis did not work at σ=0.01.**

- **p_in** regressed by +0.129 (12.108 vs 11.979). The noise is hurting in-distribution surface pressure prediction, suggesting even σ=0.01 is introducing too much perturbation to the backbone's learned representations for accurate pressure reconstruction.
- **p_oodc** regressed by +0.157. No OOD benefit — the noise did not help the model generalize to unseen conditions.
- **p_re** regressed most significantly at +0.250. Reynolds number extrapolation got worse, not better.
- **p_tan** regressed by +0.109. Tandem transfer also suffered.

**Why it failed:**

1. **The noise compounds across all 3 TransolverBlocks.** Each block adds independent noise, and the downstream blocks receive noisy inputs from the previous block. The effective noise at the final representation is larger than σ=0.01 — roughly √3 × 0.01 ≈ 0.017 for uncorrelated additive noise, but the actual effect is amplified by the attention and MLP transformations between blocks.

2. **The EMA model benefits are limited.** While the EMA model doesn't inject noise at eval time, it averages parameters trained with noise — which means the learned weights may be suboptimal for clean (noise-free) inference.

3. **The baseline is already well-regularized.** With weight decay, data augmentation (aoa_perturb, dsdf2_sigma, gap_stagger), and EMA, the model already has strong regularization. Adding hidden noise on top provides no marginal benefit and slightly degrades the signal-to-noise ratio in the learned representations.

4. **σ=0.01 may be either too large (hurts in-dist) or too small (no OOD benefit) — a lose-lose calibration point.** The regression in ALL metrics suggests it's too large. However, making it smaller would likely have zero effect.

### Research notes

The researcher agent confirmed that hidden noise injection is theoretically equivalent to Tikhonov regularization / Sobolev-norm penalty on the learned mapping (Camuto et al., NeurIPS 2020). However, the theory assumes the noise is the dominant regularizer — in our heavily-regularized baseline, it is redundant. The fact that spectral norm on SRF heads (#2243) improved p_in while backbone noise hurt it suggests the OOD problem in this model is more nuanced than "backbone representations are erratic" — it may be about specific failure modes (e.g., attention pattern collapse on OOD inputs) rather than general smoothness.

### Suggested follow-ups
- **Noise only on last TransolverBlock** — reduce compounding effect, target the representation closest to the output heads.
- **Scheduled noise (anneal σ from 0.01 to 0 over training)** — use noise as early-training regularizer, remove it for fine-tuning phase.
- **Multiplicative noise (dropout-like)** instead of additive — scales with activation magnitude, may be better calibrated.
- **Attention-targeted perturbation** — add noise to the attention logits rather than the hidden states, directly targeting the attention pattern stability on OOD inputs.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2254 because ADVISOR: **Closing — all 4 metrics regress on 2-seed average.**

| Metric | Baseline | 2-seed Avg | Δ |
|--------|----------|-----------|---|
| p_in | 11.979 | 12.108 | +1.1% ✗ |
| p_oodc | 7.643 | 7.826 | +2.4% ✗ |
| p_tan | 28.341 | 28.451 | +0.4% ✗ |
| p_re | 6.300 | 6.541 | +3.8% ✗ |

W&B runs: qhex5dph (s42), sq588kij (s73) — verified, match your reported values.

**Analysis:** The baseline's heavy regularization stack (EMA 0.999 + cosine LR + weight decay + Lion) already prevents the backbone from developing brittle representations. Adding σ=0.01 Gaussian noise across all 3 TransolverBlocks compounds through the residual connections and introduces training instability (high seed variance on p_oodc: 8.03 vs 7.62). The backbone OOD failure identified by spectral norm experiments is real, but additive noise at this level is too blunt an instrument.

**Key insight:** Backbone OOD fragility exists but needs targeted intervention (e.g., attention-level perturbation) rather than hidden-state noise. Your suggestion of attention-targeted perturbation is the most mechanistically motivated follow-up — I'll consider it for a future assignment.

Good work on the thorough analysis, edward. Moving you to a fresh experiment..
